### PR TITLE
[WIP] New Plugin Life-cycle API 

### PR DESCRIPTION
**Pull request by JRoy**
WIP PR for the new plugin life-cycle API.

See #4317 for a better description

### CODE REVIEW

File: Spigot-API-Patches/0229-Add-Server-isStarting-Bukkit-isStarting-API.patch [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
+index 7c0a788900c93c29d14d8c45ac5ae3317cf4a94e..56071c2ef5f1a3a831492bf16be4b966b3629efd 100644
+--- a/src/main/java/org/bukkit/Server.java
++++ b/src/main/java/org/bukkit/Server.java
+@@ -1548,6 +1548,13 @@ public interface Server extends PluginMessageRecipient {
+      */
+     boolean isStopping();
+ 
++    /**
++     * Checks if the server is in the process of being started for the first time.
```
Comment (kashike) [inline Spigot-API-Patches/0229-Add-Server-isStarting-Bukkit-isStarting-API.patch]:
```suggestion
+     * Checks if the server is in the process of being started.
```

File: Spigot-API-Patches/0229-Add-Server-isStarting-Bukkit-isStarting-API.patch [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
+index bfe842364ee0a4bf39dacdbb6972477d57a4ef8a..ec99b2d63931857cedfb36900e6de2f0f36e2869 100644
+--- a/src/main/java/org/bukkit/Bukkit.java
++++ b/src/main/java/org/bukkit/Bukkit.java
+@@ -1772,6 +1772,15 @@ public final class Bukkit {
+         return server.isStopping();
+     }
+ 
++    /**
++     * Checks if the server is in the process of being started for the first time.
```
Comment (kashike) [inline Spigot-API-Patches/0229-Add-Server-isStarting-Bukkit-isStarting-API.patch]:
```suggestion
+     * Checks if the server is in the process of being started.
```

**Review by kashike — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0232-Add-plugin-reload-policy.patch [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
++     * <p>
++     * This method will induce the following behavior:
++     * <ul>
++     * <li>Reloads all configs
++     * <li>Reloads server icon
++     * <li>Reloads ban lists
++     * <li>Reloads all plugins that which both their dependencies and themselves meet one the following:
++     * <ul>
++     * <li>Have the {@link org.bukkit.plugin.PluginReloadPolicy#ALWAYS ALWAYS} reload policy.
++     * <li>Have the {@link org.bukkit.plugin.PluginReloadPolicy#SAME_HASH SAME_HASH} reload policy and the plugin's file is unmodified from startup.
```
Comment (Proximyst) [inline Spigot-API-Patches/0232-Add-plugin-reload-policy.patch:27]: *cries in `</li>`*

File: Spigot-API-Patches/0232-Add-plugin-reload-policy.patch [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
++
++    /**
++     * Looks up a PluginReloadPolicy by possible plugin.yml values.
++     *
++     * @param name Name of the policy.
++     * @return Specified value, or null if not found.
++     */
++    public static PluginReloadPolicy getByName(@NotNull String name) {
++        try {
++            return valueOf(name.toUpperCase(Locale.ENGLISH).replaceAll("[^A-Z-]", "").replace("-", "_"));
```
Comment (Proximyst) [inline Spigot-API-Patches/0232-Add-plugin-reload-policy.patch:183]:
```suggestion
+            return valueOf(name.toUpperCase(Locale.ENGLISH).replace("-", "_"));
```

Don't correct the user; them entering e.g. `123` is not grounds for Paper to care as that's their fault. The replacing `-` -> `_` is however alright as that's a very easy thing to fuck up (*cough* Bungee & depend/-s...).

**Review by Proximyst — CHANGES_REQUESTED** [review]

**Comment by JRoy** [comment]
being continued by @Owen1212055 in #8108

**Pull request closed without merging**